staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go

```diff
@@ -165,6 +212,121 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+	regionBeta := TopologyKeyExist(v1.LabelFailureDomainBetaRegion, pv.Spec.NodeAffinity)
+	if zoneBeta || regionBeta {
+		// Beta NodeAffinity exists, GA NodeAffinity not exist
+		return false, v1.LabelFailureDomainBetaZone, v1.LabelFailureDomainBetaRegion
+	}
+
+	// If nothing is in the NodeAfinity, we should check pv labels
+	_, zoneGA = pv.Labels[v1.LabelTopologyZone]
+	_, regionGA = pv.Labels[v1.LabelTopologyRegion]
+	if zoneGA || regionGA {
+		// NodeAffinity not exist, GA label exists
+		return true, v1.LabelTopologyZone, v1.LabelTopologyRegion
+	}
+
+	// If GA label not exists, check beta version
+	_, zoneBeta = pv.Labels[v1.LabelFailureDomainBetaZone]
+	_, regionBeta = pv.Labels[v1.LabelFailureDomainBetaRegion]
+	if zoneBeta || regionBeta {
+		// Beta label exists, NodeAffinity not exist, GA label not exists
+		return false, v1.LabelFailureDomainBetaZone, v1.LabelFailureDomainBetaRegion
+	}
+
+	// No labels or NodeAffinity exist, default to GA version
+	return true, v1.LabelTopologyZone, v1.LabelTopologyRegion
+}
+
+// TopologyKeyExist checks if a certain key exists in a VolumeNodeAffinity
+func TopologyKeyExist(key string, vna *v1.VolumeNodeAffinity) bool {
+	if vna == nil || vna.Required == nil || vna.Required.NodeSelectorTerms == nil || len(vna.Required.NodeSelectorTerms) == 0 {
+		return false
+	}
+
+	for _, nodeSelectorTerms := range vna.Required.NodeSelectorTerms {
+		nsrequirements := nodeSelectorTerms.MatchExpressions
+		for _, nodeSelectorRequirement := range nsrequirements {
+			if nodeSelectorRequirement.Key == key {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// translateTopologyFromCSIToInTree
+// 1. Remove any existing CSI topologyKey from NodeAffinity
+// 2. Add Kubernetes Topology in the NodeAffinity if it does not exist
+// 3. Add Kubernetes Topology labels
+func translateTopologyFromCSIToInTree(pv *v1.PersistentVolume, topologyKey string) error {
+
+	csiTopologyZoneValues := getTopologyValues(pv, topologyKey)
+	csiTopologyZoneValStr := strings.Join(csiTopologyZoneValues, labelMultiZoneDelimiter)
+
+	// 1. Frist remove the CSI topology Key
+	removeTopology(pv, topologyKey)
+
+	_, zoneLabel, regionLabel := getTopologyLabel(pv)
+	zoneLabelVal, zoneOK := pv.Labels[zoneLabel]
+	regionLabelVal, regionOK := pv.Labels[regionLabel]
+
+	// 2. Add Kubernetes Topology in the NodeAffinity if it does not exist
+	// Check if Kubernetes Topology already exist
+	topologyZoneValue := getTopologyValues(pv, zoneLabel)
+	if len(topologyZoneValue) == 0 {
+		// No Kubernetes Topology exist in the current PV, we need to add it
+		if zoneOK {
+			zones := strings.Split(zoneLabelVal, labelMultiZoneDelimiter)
+			if len(zones) > 0 {
+				err := addTopology(pv, zoneLabel, zones)
+				if err != nil {
+					return fmt.Errorf("Failed to add Kubernetes topology zone to PV NodeAffinity. %v", err)
+				}
+			}
+		} else {
+			// nothing is in the label, we try our best to add the CSI topology zone in
```

Comment:
I think we should prioritize CSI first before the PV label. Someone would have to manually create their own PV or have their own CSI driver to get into the scenario where a CSI PV object would have a intree label on it.

Reply:
See comments above.

Reply:
Sorry which comment are you referring to?

Reply:
Sorry wrong reply, lol. I thought I was replying to another comment. NVM

So here, you mean if there is a CSI topology and also PV labels. We should use the CSI topology to generate Kubernetes topology first?

Reply:
Yes, we should prefer using what's in CSI topology over PV labels because in the normal case we expect only CSI topology to exist. If PV label exists, something unusual is going on.

Reply:
Done.